File: src/langgraph_agent_blueprint/api/schemas.py
"""Pydantic DTOs for FastAPI chat, approval, streaming, and frontend status payloads."""

from __future__ import annotations

from datetime import datetime
from typing import Any
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, RootModel, field_validator

from langgraph_agent_blueprint.models import (
    AttachmentRef,
    ConfigDiagnostic,
    ConfigSource,
    ConfigValueOrigin,
    PermissionRequest,
)
from langgraph_agent_blueprint.utils.ids import validate_session_id, validate_thread_id

ModelIntelligenceLevel = Literal["low", "medium", "high", "very_high"]
PermissionMode = Literal["default", "accept_edits", "bypass_read_only", "plan", "strict"]


class RuntimeEventDTO(BaseModel):
    """Stable frontend event envelope emitted by graph runtime adapters."""

    id: str
    type: str
    timestamp: datetime
    session_id: str
    node: str | None = None
    severity: Literal["info", "warning", "error"] = "info"
    data: dict[str, Any] = Field(default_factory=dict)


class StreamFrame(BaseModel):
    """One SSE data frame from the live chat stream."""

    type: Literal["event", "done", "error"]
    event: RuntimeEventDTO | None = None
    session_id: str | None = None
    thread_id: str | None = None
    final_response: str | None = None
    error: str | None = None


class ChatRequest(BaseModel):
    """API request payload for starting or continuing a graph-backed chat turn."""

    message: str
    project_id: str | None = None
    session_id: str | None = None
    thread_id: str | None = None
    model_intelligence: ModelIntelligenceLevel | None = None
    permission_mode: PermissionMode | None = None
    attachments: list[AttachmentRef] = Field(default_factory=list)

    @field_validator("session_id")
    @classmethod
    def _validate_session_id(cls, value: str | None) -> str | None:
        return validate_session_id(value) if value is not None else None

    @field_validator("thread_id")
    @classmethod
    def _validate_thread_id(cls, value: str | None) -> str | None:
        return validate_thread_id(value) if value is not None else None


class ChatResponse(BaseModel):
    """API response payload containing session ids, final text, events, and pending approvals."""

    session_id: str
    thread_id: str
    final_response: str | None = None
    events: list[RuntimeEventDTO]
    usage: dict[str, Any] = Field(default_factory=dict)
    permission_required: PermissionRequest | None = None


class ChatCancelRequest(BaseModel):
    """API request payload for stopping an active streaming graph turn."""

    thread_id: str
    session_id: str | None = None
    reason: str | None = None

    @field_validator("session_id")
    @classmethod
    def _validate_session_id(cls, value: str | None) -> str | None:
        return validate_session_id(value) if value is not None else None

    @field_validator("thread_id")
    @classmethod
    def _validate_thread_id(cls, value: str) -> str:
        return validate_thread_id(value)


class ChatCancelResponse(BaseModel):
    """API response payload returned after a stop-generation request."""

    cancelled: bool
    thread_id: str
    session_id: str | None = None
    reason: str | None = None


class WorkspaceAddRequest(BaseModel):
    """Request to register and select a validated local workspace root."""

    root_path: str


class WorkspaceSelectRequest(BaseModel):
    """Request to make an existing workspace the active workspace."""

    project_id: str


class WorkspaceCheckoutRequest(BaseModel):
    """Request to checkout an existing local branch in a workspace."""

    branch: str
    confirm_dirty: bool = False


class PermissionDecisionDTO(BaseModel):
    """Explicit frontend approval/rejection decision for one pending tool call."""

    tool_call_id: str
    decision: Literal["approved", "rejected"]
    reason: str | None = None
    remember: bool = False


class LegacyPermissionDecisionDTO(BaseModel):
    """Backwards-compatible approval shape used by the initial React CLI shell."""

    approved: bool
    reason: str | None = None
    remember: bool = False


class ApprovalRequest(BaseModel):
    """API request payload used to resume an interrupted graph with a human permission decision."""

    thread_id: str
    session_id: str | None = None
    decision: PermissionDecisionDTO | LegacyPermissionDecisionDTO

    @field_validator("session_id")
    @classmethod
    def _validate_session_id(cls, value: str | None) -> str | None:
        return validate_session_id(value) if value is not None else None

    @field_validator("thread_id")
    @classmethod
    def _validate_thread_id(cls, value: str) -> str:
        return validate_thread_id(value)

    def decision_payload(self) -> dict[str, Any]:
        """Return the graph resume payload while preserving legacy approved-bool support."""

        return self.decision.model_dump(mode="json", exclude_none=True)


class RegistryItemDTO(BaseModel):
    """Common frontend metadata row for commands, skills, tools, and extension registries."""

    model_config = ConfigDict(extra="allow")

    name: str
    description: str = ""
    type: str | None = None
    status: str | None = None
    plugin_name: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class CommandRegistryDTO(RootModel[dict[str, RegistryItemDTO]]):
    """Slash-command registry response keyed by command name for backwards compatibility."""


class SkillRegistryDTO(RootModel[dict[str, RegistryItemDTO]]):
    """Skill registry response keyed by skill name for backwards compatibility."""


class ToolRegistryDTO(RootModel[dict[str, RegistryItemDTO]]):
    """Tool registry response keyed by model-callable tool name for backwards compatibility."""


class MessageDTO(BaseModel):
    """Frontend-safe normalized chat/session message."""

    id: str
    role: str
    content: str
    type: str | None = None
    tool_calls: list[dict[str, Any]] = Field(default_factory=list)
    tool_call_id: str | None = None


class ToolCallRecordDTO(BaseModel):
    """Frontend-safe tool execution record from session persistence."""

    id: str
    name: str
    status: str
    content: str = ""
    output_ref: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)
    error: dict[str, Any] | str | None = None


class ContextStateDTO(BaseModel):
    """Resolved context state for a session side panel."""

    references: list[dict[str, Any]] = Field(default_factory=list)
    fragments: list[dict[str, Any]] = Field(default_factory=list)
    attachments: list[dict[str, Any]] = Field(default_factory=list)
    budget: dict[str, Any] = Field(default_factory=dict)
    errors: list[dict[str, Any]] = Field(default_factory=list)


class ChildRunListItemDTO(BaseModel):
    """Summary row for one persisted child/subagent run."""

    child_run_id: str
    parent_session_id: str | None = None
    child_session_id: str | None = None
    child_thread_id: str | None = None
    name: str | None = None
    purpose: str | None = None
    status: str | None = None
    started_at: str | None = None
    completed_at: str | None = None
    summary: str | None = None


class ChildRunDetailDTO(BaseModel):
    """Frontend-safe child/subagent run detail."""

    metadata: dict[str, Any]
    result: dict[str, Any] = Field(default_factory=dict)
    events: list[RuntimeEventDTO] = Field(default_factory=list)


class SessionListItemDTO(BaseModel):
    """Frontend session list item with counts instead of raw storage shape."""

    session_id: str
    title: str | None = None
    created_at: str | None = None
    updated_at: str | None = None
    provider: str | None = None
    model: str | None = None
    message_count: int = 0
    event_count: int = 0
    tool_call_count: int = 0
    child_run_count: int = 0
    usage: dict[str, Any] = Field(default_factory=dict)


class SessionDetailDTO(BaseModel):
    """Frontend-safe session detail DTO."""

    session_id: str
    title: str | None = None
    created_at: str | None = None
    updated_at: str | None = None
    provider: str | None = None
    model: str | None = None
    messages: list[MessageDTO] = Field(default_factory=list)
    events: list[RuntimeEventDTO] = Field(default_factory=list)
    tool_calls: list[ToolCallRecordDTO] = Field(default_factory=list)
    todos: list[dict[str, Any]] = Field(default_factory=list)
    memory: dict[str, Any] = Field(default_factory=dict)
    usage: dict[str, Any] = Field(default_factory=dict)
    context: ContextStateDTO = Field(default_factory=ContextStateDTO)
    child_runs: list[ChildRunListItemDTO] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)


class ExportRequest(BaseModel):
    """Request body for session transcript export."""

    format: Literal["markdown", "text"] = "markdown"


class ExportRecordDTO(BaseModel):
    """Frontend-safe export result."""

    session_id: str
    format: str
    path: str
    bytes: int


class ConfigShowDTO(BaseModel):
    """Redacted effective config for read-only frontend display."""

    values: dict[str, Any]


class ConfigExplainDTO(BaseModel):
    """Config source and value-origin explanation."""

    sources: list[ConfigSource] = Field(default_factory=list)
    values: list[ConfigValueOrigin] = Field(default_factory=list)
    diagnostics: list[ConfigDiagnostic] = Field(default_factory=list)


class ConfigValidateDTO(BaseModel):
    """Config validation diagnostics."""

    ok: bool
    diagnostics: list[ConfigDiagnostic] = Field(default_factory=list)


class PluginStatusDTO(BaseModel):
    """Read-only plugin status and diagnostics."""

    plugins: list[dict[str, Any]] = Field(default_factory=list)
    contributions: list[dict[str, Any]] = Field(default_factory=list)
    errors: list[dict[str, Any]] = Field(default_factory=list)
    warnings: list[dict[str, Any]] = Field(default_factory=list)


class HookStatusDTO(BaseModel):
    """Read-only hook registry status."""

    hooks: list[dict[str, Any]] = Field(default_factory=list)


class MCPStatusDTO(BaseModel):
    """Read-only MCP discovery/status snapshot."""

    servers: list[dict[str, Any]] = Field(default_factory=list)
    tools: dict[str, dict[str, Any]] = Field(default_factory=dict)
    resources: dict[str, list[dict[str, Any]]] = Field(default_factory=dict)
    prompts: dict[str, list[dict[str, Any]]] = Field(default_factory=dict)
    events: list[dict[str, Any]] = Field(default_factory=list)
    transport_support: dict[str, bool] = Field(default_factory=dict)
    invalid_servers: list[dict[str, Any]] = Field(default_factory=list)
    warnings: list[dict[str, Any]] = Field(default_factory=list)
    errors: list[dict[str, Any]] = Field(default_factory=list)


class ObservabilityStatusDTO(BaseModel):
    """Read-only observability backend status."""

    enabled: bool
    mode: str
    sdk_installed: bool | None = None
    base_url_configured: bool = False
    public_key_present: bool = False
    secret_key_present: bool = False
    environment: str | None = None
    release: str | None = None
    capture_inputs: bool = False
    capture_outputs: bool = False
    runtime_events_mode: str | None = None
    auth_check: str | None = None
    last_error: str | None = None

File: frontend/src/api/schemas.ts
export type RuntimeEventSeverity = "info" | "warning" | "error";
export type AgentActivityStatus = "pending" | "running" | "success" | "error" | "blocked";

export type AgentActivitySource = {
  kind: string;
  name?: string | null;
  component?: string | null;
};

export type AgentActivityRef = {
  kind: string;
  path?: string | null;
  name?: string | null;
  id?: string | null;
};

export type AgentActivityEvent = {
  id: string;
  type: string;
  source: AgentActivitySource;
  category: string;
  status?: AgentActivityStatus | null;
  title: string;
  summary?: string | null;
  data: Record<string, unknown>;
  refs: AgentActivityRef[];
};

export type RuntimeStreamEvent =
  | { kind: "assistant_delta"; message_id: string; delta: string }
  | { kind: "assistant_final"; message_id: string; content: string }
  | { kind: "progress"; message: string; stage?: string | null; message_id?: string | null }
  | {
      kind: "tool_lifecycle";
      phase: "scheduled" | "started" | "completed" | "failed" | "blocked" | "permission_required";
      tool_call_id: string;
      tool_name: string;
      title?: string | null;
      args_summary?: string | null;
      result_summary?: string | null;
      command?: string | null;
      path?: string | null;
      exit_code?: number | null;
      duration_ms?: number | null;
      error?: { type: string; message: string } | null;
      details?: Record<string, unknown>;
    }
  | {
      kind: "permission_state";
      status: "required" | "approved" | "rejected" | "blocked";
      tool_call_id: string;
      tool_name: string;
      action?: string | null;
      risk?: string | null;
      args_summary?: string | null;
      reason?: string | null;
      args?: Record<string, unknown>;
      scope?: "tool" | "subagent";
      parent_session_id?: string | null;
      parent_thread_id?: string | null;
      child_session_id?: string | null;
      child_thread_id?: string | null;
      child_run_id?: string | null;
      subagent_name?: string | null;
    }
  | { kind: "error"; message: string; error_type?: string | null; recoverable?: boolean | null }
  | { kind: "artifact"; artifact_id: string; artifact_kind: string; uri?: string | null; title?: string | null; metadata?: Record<string, unknown> };

export type RuntimeEvent = {
  id: string;
  type: string;
  timestamp: string;
  session_id: string;
  node?: string | null;
  severity: RuntimeEventSeverity;
  data: Record<string, unknown>;
};

export type AttachmentRef = {
  id: string;
  kind: string;
  name?: string | null;
  uri?: string | null;
  path?: string | null;
  mime_type?: string | null;
  size_bytes?: number | null;
  trust?: string;
  metadata?: Record<string, unknown>;
};

export type PermissionRequest = {
  type?: string;
  tool_call_id: string;
  tool_name: string;
  action?: string | null;
  risk?: string | null;
  args_summary?: string | null;
  reason?: string | null;
  args?: Record<string, unknown>;
  scope?: "tool" | "subagent";
  parent_session_id?: string | null;
  parent_thread_id?: string | null;
  child_session_id?: string | null;
  child_thread_id?: string | null;
  child_run_id?: string | null;
  subagent_name?: string | null;
};

export type PermissionDecisionDTO = {
  tool_call_id: string;
  decision: "approved" | "rejected";
  reason?: string | null;
};

export type ModelIntelligenceLevel = "low" | "medium" | "high" | "very_high";
export type PermissionMode = "default" | "accept_edits" | "bypass_read_only" | "plan" | "strict";

export type ChatRequest = {
  message: string;
  project_id?: string | null;
  session_id?: string | null;
  thread_id?: string | null;
  model_intelligence?: ModelIntelligenceLevel | null;
  permission_mode?: PermissionMode | null;
  attachments?: AttachmentRef[];
};

export type GitStatusSummary = {
  staged: number;
  unstaged: number;
  untracked: number;
  conflicted: number;
};

export type WorkspaceInfo = {
  project_id: string;
  display_name: string;
  root_path: string;
  is_git_repo: boolean;
  current_branch?: string | null;
  branches: string[];
  git_status?: GitStatusSummary | null;
  dirty: boolean;
  created_at?: string | null;
  last_opened_at?: string | null;
};

export type WorkspaceCheckoutRequest = {
  branch: string;
  confirm_dirty?: boolean;
};

export type WorkspaceCheckoutResult = {
  ok: boolean;
  workspace: WorkspaceInfo;
  message: string;
};

export type ChatResponse = {
  session_id: string;
  thread_id: string;
  conversation_id?: string | null;
  final_response?: string | null;
  events: RuntimeEvent[];
  usage?: Record<string, unknown>;
  permission_required?: PermissionRequest | null;
};

export type ChatCancelRequest = {
  thread_id: string;
  session_id?: string | null;
  reason?: string | null;
};

export type ChatCancelResponse = {
  cancelled: boolean;
  thread_id: string;
  session_id?: string | null;
  reason?: string | null;
};

export type ConversationStatus = "active" | "archived" | "deleted";

export type ConversationRecordDTO = {
  conversation_id: string;
  session_id: string;
  thread_id: string;
  user_id: string;
  title?: string | null;
  status: ConversationStatus;
  created_at: string;
  updated_at: string;
  metadata: Record<string, unknown>;
  schema_version: number;
};

export type ConversationListItemDTO = {
  conversation_id: string;
  session_id: string;
  thread_id: string;
  title?: string | null;
  status: ConversationStatus;
  message_count: number;
  event_count: number;
  tool_call_count: number;
  artifact_count: number;
  created_at: string;
  updated_at: string;
};

export type ConversationMessageDTO = {
  message_id: string;
  role: string;
  content: string;
  created_at: string;
  metadata?: Record<string, unknown>;
};

export type ConversationEventDTO = {
  event_id: string;
  conversation_id: string;
  type: string;
  payload: Record<string, unknown>;
  metadata: Record<string, unknown>;
  created_at: string;
};

export type ConversationToolCallDTO = {
  tool_call_id: string;
  name: string;
  status: string;
  metadata: Record<string, unknown>;
};

export type ConversationArtifactDTO = {
  artifact_id: string;
  kind: string;
  uri?: string | null;
  metadata: Record<string, unknown>;
};

export type ConversationDetailDTO = {
  conversation: ConversationRecordDTO;
  messages: ConversationMessageDTO[];
  events: ConversationEventDTO[];
  tool_calls: ConversationToolCallDTO[];
  artifacts: ConversationArtifactDTO[];
};

export type StreamFrame =
  | { type: "event"; event: RuntimeEvent }
  | { type: "done"; session_id?: string | null; thread_id?: string | null; conversation_id?: string | null; final_response?: string | null }
  | { type: "error"; error: string };

export type ApprovalRequest = {
  thread_id: string;
  session_id?: string | null;
  decision: PermissionDecisionDTO;
};

export type RegistryItem = {
  name: string;
  description?: string;
  type?: string | null;
  status?: string | null;
  plugin_name?: string | null;
  metadata?: Record<string, unknown>;
  [key: string]: unknown;
};

export type RegistryMap = Record<string, RegistryItem>;

export type MessageDTO = {
  id: string;
  role: string;
  content: string;
  type?: string | null;
  tool_calls?: Record<string, unknown>[];
  tool_call_id?: string | null;
};

export type ToolCallRecordDTO = {
  id: string;
  name: string;
  status: string;
  content?: string;
  output_ref?: string | null;
  metadata?: Record<string, unknown>;
  error?: Record<string, unknown> | string | null;
};

export type ContextStateDTO = {
  references: Record<string, unknown>[];
  fragments: Record<string, unknown>[];
  attachments: Record<string, unknown>[];
  budget: Record<string, unknown>;
  model_context?: Record<string, unknown>;
  errors: Record<string, unknown>[];
};

export type SessionListItemDTO = {
  session_id: string;
  title?: string | null;
  created_at?: string | null;
  updated_at?: string | null;
  provider?: string | null;
  model?: string | null;
  message_count: number;
  event_count: number;
  tool_call_count: number;
  child_run_count: number;
  usage?: Record<string, unknown>;
};

export type ChildRunListItemDTO = {
  child_run_id: string;
  parent_session_id?: string | null;
  child_session_id?: string | null;
  child_thread_id?: string | null;
  name?: string | null;
  purpose?: string | null;
  status?: string | null;
  started_at?: string | null;
  completed_at?: string | null;
  summary?: string | null;
};

export type SessionDetailDTO = {
  session_id: string;
  title?: string | null;
  created_at?: string | null;
  updated_at?: string | null;
  provider?: string | null;
  model?: string | null;
  messages: MessageDTO[];
  events: RuntimeEvent[];
  tool_calls: ToolCallRecordDTO[];
  todos: Record<string, unknown>[];
  memory: Record<string, unknown>;
  usage: Record<string, unknown>;
  context: ContextStateDTO;
  child_runs: ChildRunListItemDTO[];
  metadata: Record<string, unknown>;
};

export type ConfigShowDTO = { values: Record<string, unknown> };
export type ConfigExplainDTO = {
  sources: Record<string, unknown>[];
  values: Record<string, unknown>[];
  diagnostics: Record<string, unknown>[];
};
export type ConfigValidateDTO = { ok: boolean; diagnostics: Record<string, unknown>[] };
export type PluginStatusDTO = {
  plugins: Record<string, unknown>[];
  contributions: Record<string, unknown>[];
  errors: Record<string, unknown>[];
  warnings: Record<string, unknown>[];
};
export type HookStatusDTO = { hooks: Record<string, unknown>[] };
export type MCPStatusDTO = {
  servers: Record<string, unknown>[];
  tools: Record<string, Record<string, unknown>>;
  resources: Record<string, Record<string, unknown>[]>;
  prompts: Record<string, Record<string, unknown>[]>;
  invalid_servers: Record<string, unknown>[];
  warnings: Record<string, unknown>[];
  errors: Record<string, unknown>[];
};
export type ObservabilityStatusDTO = {
  enabled: boolean;
  mode: string;
  sdk_installed?: boolean | null;
  base_url_configured?: boolean;
  public_key_present?: boolean;
  secret_key_present?: boolean;
  environment?: string | null;
  release?: string | null;
  capture_inputs?: boolean;
  capture_outputs?: boolean;
  runtime_events_mode?: string;
  last_error?: string | null;
};
